sopht_mpi/utils/mpi_utils_3d.py

```diff
@@ -98,188 +124,1003 @@ def __init__(self, ghost_size, mpi_construct):
+                self.ghost_size,
+                self.ghost_size,
+            ],
         )
-        self.send_next_along_z_type.Commit()
-        self.recv_next_along_z_type = mpi_construct.dtype_generator.Create_subarray(
-            sizes=self.field_size,
+        self.send_to_pz_0_0_type.Commit()
+        self.recv_from_nz_0_0_type = self.mpi_construct.dtype_generator.Create_subarray(
+            sizes=self.field_size_with_ghost,
             subsizes=[self.ghost_size, self.field_size[1], self.field_size[2]],
-            starts=[self.field_size[0] - self.ghost_size, 0, 0],
+            starts=[0, self.ghost_size, self.ghost_size],
         )
-        self.recv_next_along_z_type.Commit()
-        # Along Z (prev)
-        self.send_previous_along_z_type = mpi_construct.dtype_generator.Create_subarray(
-            sizes=self.field_size,
+        self.recv_from_nz_0_0_type.Commit()
+        # Comm. along -Z direction: send to (-1, 0, 0) with recv from (+1, 0, 0) block
+        self.send_to_nz_0_0_type = self.mpi_construct.dtype_generator.Create_subarray(
+            sizes=self.field_size_with_ghost,
             subsizes=[self.ghost_size, self.field_size[1], self.field_size[2]],
-            starts=[self.ghost_size, 0, 0],
+            starts=[self.ghost_size, self.ghost_size, self.ghost_size],
         )
-        self.send_previous_along_z_type.Commit()
-        self.recv_previous_along_z_type = mpi_construct.dtype_generator.Create_subarray(
-            sizes=self.field_size,
+        self.send_to_nz_0_0_type.Commit()
+        self.recv_from_pz_0_0_type = self.mpi_construct.dtype_generator.Create_subarray(
+            sizes=self.field_size_with_ghost,
             subsizes=[self.ghost_size, self.field_size[1], self.field_size[2]],
-            starts=[0, 0, 0],
+            starts=[
+                self.field_size_with_ghost[0] - self.ghost_size,
+                self.ghost_size,
+                self.ghost_size,
+            ],
         )
-        self.recv_previous_along_z_type.Commit()
+        self.recv_from_pz_0_0_type.Commit()
 
-        # Initialize requests list for non-blocking comm
-        self.comm_requests = []
+        if self.full_exchange:
+            # (2) For edges
+            # Since there are 12 edges on a cube, each with a send and recv datatypes,
+            # we need 12 * 2 = 24 datatypes
+            # Comm. along +Y, +X: send to (0, +1, +1) with recv from (0, -1, -1) block
+            self.send_to_0_py_px_type = (
+                self.mpi_construct.dtype_generator.Create_subarray(
+                    sizes=self.field_size_with_ghost,
+                    subsizes=[self.field_size[0], self.ghost_size, self.ghost_size],
+                    starts=[
+                        self.ghost_size,
+                        self.field_size_with_ghost[1] - 2 * self.ghost_size,
+                        self.field_size_with_ghost[2] - 2 * self.ghost_size,
+                    ],
+                )
```

Comment:
From here on to line 737 there is so much code repeat! Is there a way to avoid this? Can you check how people do communication in MPI-based solvers or packages, or even PPM for instance? The more code you write, the more to maintain, the more chances for bugs to pop up...